### ISSUE TITLE
Is it possible to set a custom location for the 'parinfer-rust' folder containing the .so file?

### ISSUE THREAD

**Issue by johanthoren** [justinbarclay/parinfer-rust-mode#35]
To make parinfer-rust-mode integrate better with doom-emacs it would be useful to be able to specify where the .so file should be downloaded (in this case ~/.emacs.d/.local/etc/parinfer-rust). Is there a way to set this before downloading and installing it and have parinfer-rust-mode look for it in said location?

**Comment by justinbarclay** [comment]
Definitely!
If you do `(setq parinfer-rust-library "/path/to/library.so")`, it will search for the library there and failing that ask to download the library to that location.
https://github.com/justinbarclay/parinfer-rust-mode#customizations

**Comment by johanthoren** [comment]
@justinbarclay, thanks a lot for the hint! I apologize for missing that. I'm hoping you could perhaps offer some further guidance. I'm trying the following snippet:
```emacs-lisp
(use-package! parinfer-rust-mode
  :hook ((emacs-lisp-mode
          clojure-mode
          scheme-mode
          lisp-mode
          racket-mode
          hy-mode) . parinfer-rust-mode)
  :init
  (setq! parinfer-rust-library (concat user-emacs-directory
                                       ".local/etc/parinfer-rust/"
                                       parinfer-rust--lib-name))
  :config  
  (map! :map parinfer-rust-mode-map
        :localleader
        :desc "Switch parinfer-rust-mode" "m" #'parinfer-rust-switch-mode
        :desc "Temporarily disable parinfer-rust-mode" "M" #'parinfer-rust-switch-mode))
```
But it's still being downloaded to ~/.emacs.d/parinfer-rust/. What am I doing wrong?

**Comment by justinbarclay** [comment]
Ok I was able to recreate this issue:
1. Make sure remove parinfer-rust artifacts (.so file and the package itself)
2. In `init` set to non-default directory
3. load up a fresh instance of emacs
4. it download .so to `.emacs.d/parinfer-rust`

So what looks to be the culprit is that we have to define `parinfer-rust--lib-name and `parinfer-rust-library` to make the byte-compiler happy. And because these have a default value, on the initial downloading and loading of the package it uses these values for byte compiling (which executes some code to pulldown the library), which can be seen [here](https://github.com/justinbarclay/parinfer-rust-mode/blob/main/parinfer-rust-mode.el#L165).

In other test cases I've been able to delete the `parinfer-rust` folder and on next load of emacs/parinfer-rust-mode it will attempt to download to the correct directory.

At the moment, no immediate fixes are apparent to me. But I will spend sometime this weekend and think about this more.

**Comment by johanthoren** [comment]
@justinbarclay, I'm somewhat glad to hear that I wasn't going crazy! Thanks a lot for your feedback and fast responses!

**Comment by justinbarclay** [comment]
Hey, @johanthoren, Doom uses straight under the hood right? Can you try pointing it to [defer-download](https://github.com/justinbarclay/parinfer-rust-mode/tree/defer-download) and see if it starts respecting the preferred library location?

**Comment by johanthoren** [comment]
@justinbarclay, it does, and using that branch it seems to work if I define `parinfer-rust--lib-name` before using `use-package!`. Why is it that the name for gnu/linux and windows are defined twice [here](https://github.com/justinbarclay/parinfer-rust-mode/blob/251883ab238eb0b131c7b0b22d7307a6ed281396/parinfer-rust-mode.el#L159), but only once for darwin?

**Comment by justinbarclay** [comment]
Oh thanks for calling that out! I don't think there is a good reason for them to be defined twice.

Hmm, can you confirm that you've tried setting the location and filename with `parinfer-rust-library` doesn't work for you? Something like: 
```(setq parinfer-rust-library (concat user-emacs-directory ".local/etc/parinfer-rust/parinfer-rust.so"))```

I'm getting familiar with my code again and `parinfer-rust--lib-name` should be a private/uncustomizeable value (it's defined as a `defconst`). That is why you need to define it before your `use-package` reference, because if you tried to define it after it's getting set as immutable.

As an FYI it's static because `parinfer-rust--lib-name` defines, at minimum, what artifact to pull down from github. Meaning, that If the user could change it at library runtime it might fail to pull down anything.

**Comment by johanthoren** [comment]
I can confirm that it works if I set it as you describe above. It also works when I `defvar parinfer-rust--library-name` before I do the `use-package`. That has the added benefit of giving the file the correct name depending on OS. Would it be considered a bad idea to do it this way?
```emacs-lisp
(if (not (featurep! +rust))
    (use-package! parinfer
      ...)
  ;; This needs to be defined before calling use-package! since we want to set
  ;; the correct download location.
  (defvar parinfer-rust--lib-name (cond ((eq system-type 'darwin)
                                         "parinfer-rust-darwin.so")
                                        ((eq system-type 'gnu/linux)
                                         "parinfer-rust-linux.so")
                                        ((eq system-type 'windows-nt)
                                         "parinfer-rust-windows.dll")))
  (use-package! parinfer-rust-mode
    :hook ((emacs-lisp-mode
            clojure-mode
            scheme-mode
            lisp-mode
            racket-mode
            hy-mode) . parinfer-rust-mode)
    :init
    (setq! parinfer-rust-library (concat user-emacs-directory
                                         ".local/etc/parinfer-rust/"
                                         parinfer-rust--lib-name))
    :config
    (map! :map parinfer-rust-mode-map
          :localleader
          :desc "Switch parinfer-rust-mode" "m" #'parinfer-rust-switch-mode
          :desc "Temporarily disable parinfer-rust-mode" "M"
                #'parinfer-rust-switch-mode)))
```
Would it be better to name the variable something else entirely?

**Comment by justinbarclay** [comment]
Awesome, we're making progress 😃 

Ahh right, you're doing this so that it works across OSes. Yeah, this approach makes sense then.

Can I ask you to change the defvar to `parinfer-rust--lib-name` to something else, maybe `doom-parinfer-rust-lib`? I have some small concern with defining a private variable outside of the library where it's used in ( Due to elisp lacking namespaces, a `--` denotes a name that is meant to be private to that package 🤕 ). In the current context I can't think of where this would actually cause problems but I'd rather be safe than sorry. 😄

**Comment by johanthoren** [comment]
Thanks for the input. I'll change it to something like that. Elisp is not my usual dialect, so thanks for pointing it out.

I will wait for you to merge this fix with the master branch before pinning to that commit. Thanks a lot!

**Comment by justinbarclay** [comment]
This has been merged into main now 😄 
Thanks for your help with this!
